File: templates/llm-output-r-eval-parse-detector/examples/good.R
# Good fixture: same intents as bad.R, expressed without eval(parse(text=...)).

# 1: dynamic variable -> use a list (the idiomatic R answer)
models <- list()
for (i in 1:5) {
  models[[paste0("model_", i)]] <- lm(y ~ x, data = df)
}

# 2: dynamic formula -> as.formula() then standard call
build_and_run <- function(col) {
  f <- as.formula(paste0("y ~ ", col))
  summary(lm(f, data = df))
}

# 3: dynamic column reference -> [[ ]] indexing, no parsing
get_value <- function(name) {
  df[[name]]
}

# 4: parse(file = ...) is a legitimate "source another script" pattern
#    — the detector must NOT flag this.
exprs <- parse(file = "helpers.R")
eval(exprs)

# 5: evaluating an already-quoted expression (normal metaprogramming)
e <- quote(x + 1)
eval(e)

# 6: bquote / substitute for code generation
template <- bquote(z <- .(a) + .(b), list(a = 1, b = 2))
eval(template)

# 7: a comment that mentions eval(parse(text = ...)) must not trigger:
# avoid eval(parse(text = ...)) — it's an injection sink

# 8: a string literal that mentions it must not trigger either:
warning_message <- "do not use eval(parse(text = x)) in production"

# 9: an audited internal use can be suppressed inline:
internal_eval <- function(s) {
  eval(parse(text = s)) # eval-parse-ok: knitr chunk option parser, internal-only
}

File: templates/llm-output-clojure-eval-read-string-detector/examples/good.clj
;; Good fixture: same intents as bad.clj, expressed without (eval (read-string ...)).

;; 1: dynamic "var name" -> use a map keyed by keyword (the idiomatic answer)
(def models
  (into {}
        (for [i (range 5)]
          [(keyword (str "model-" i)) (fit data)])))

;; 2: dynamic column reference -> just use get / keyword lookup
(defn get-value [name]
  (get row (keyword name)))

;; 3: parsing untrusted input -> use clojure.edn/read-string (data only,
;;    no code evaluation). The detector must NOT flag this.
(require '[clojure.edn :as edn])
(def config (edn/read-string (slurp "config.edn")))

;; 4: read-string alone (without eval) returns a form (data). On its own
;;    it's harmless — and the detector must NOT flag it.
(def parsed-form (read-string "(+ 1 2)"))
(println parsed-form) ; => (+ 1 2)

;; 5: normal metaprogramming — eval of a quoted/syntax-quoted form
;;    (NOT a string). The detector must NOT flag this.
(eval '(+ 1 2))
(eval `(+ ~a ~b))

;; 6: a macro is the right tool for compile-time code generation
(defmacro defn-traced [name args & body]
  `(defn ~name ~args
     (println "calling" '~name)
     ~@body))

;; 7: a comment that mentions (eval (read-string ...)) must not trigger:
;; avoid (eval (read-string s)) — it's an injection sink

;; 8: a string literal that mentions it must not trigger either:
(def warning-message "do not use (eval (read-string s)) in production")

;; 9: a regex literal that mentions it must not trigger either:
(def pattern #"(eval (read-string \"x\"))")

;; 10: an audited internal use can be suppressed inline:
(defn internal-eval [s]
  (eval (read-string s))) ;; eval-read-string-ok: REPL-helper macro, internal-only
